jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c

```diff
@@ -1687,6 +1687,191 @@ ecma_builtin_typedarray_prototype_find_index (ecma_value_t this_arg, /**< this a
+                                             uint32_t *index, /**< pointer to the given index */
+                                             bool is_end_index) /**< true - normalize the end index */
+{
+  ecma_number_t num_var;
+  ecma_value_t ret_value = ecma_get_number (arg_value, &num_var);
+
+  if (ECMA_IS_VALUE_ERROR (ret_value))
+  {
+    return ret_value;
+  }
+
+  if (is_end_index && ecma_number_is_nan (num_var))
+  {
+    *index = length;
+  }
+  else
+  {
+    *index = ecma_builtin_helper_array_index_normalize (num_var, length);
+  }
+
+  return ECMA_VALUE_EMPTY;
+} /* ecma_builtin_get_uint32_value_from_argument */
+
+/**
+ * The %TypedArray%.prototype object's 'copyWithin' routine
+ *
+ * See also:
+ *          ECMA-262 v6, 22.2.3.5
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_typedarray_prototype_copy_within (ecma_value_t this_arg, /**< this argument */
+                                               const ecma_value_t args[], /**< arguments list */
+                                               ecma_length_t args_number) /**< number of arguments */
+{
+  if (!ecma_is_typedarray (this_arg))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is not a TypedArray."));
+  }
+
+  ecma_object_t *typedarray_p = ecma_get_object_from_value (this_arg);
+  uint32_t length = ecma_typedarray_get_length (typedarray_p);
+  lit_utf8_byte_t *typedarray_buffer_p = ecma_typedarray_get_buffer (typedarray_p);
+  uint8_t shift = ecma_typedarray_get_element_size_shift (typedarray_p);
+  uint8_t element_size = (uint8_t) (1 << shift);
+  uint32_t limit = length * element_size;
+  uint32_t target = 0;
+  uint32_t start = 0;
+  uint32_t end = limit;
+
+  if (args_number > 0)
+  {
+    ecma_value_t target_value = ecma_builtin_get_uint32_value_from_argument (args[0], length, &target, false);
+    target *= element_size;
+
+    if (ECMA_IS_VALUE_ERROR (target_value))
+    {
+      return target_value;
+    }
+
+    if (args_number > 1)
+    {
+      ecma_value_t start_value = ecma_builtin_get_uint32_value_from_argument (args[1], length, &start, false);
+      start *= element_size;
+
+      if (ECMA_IS_VALUE_ERROR (start_value))
+      {
+        return start_value;
+      }
+
+      if (args_number > 2)
+      {
+        ecma_value_t end_value = ecma_builtin_get_uint32_value_from_argument (args[2], length, &end, true);
+        end *= element_size;
+
+        if (ECMA_IS_VALUE_ERROR (end_value))
+        {
+          return end_value;
+        }
+      }
+    }
+  }
+
+  int32_t distance = (int32_t) (end - start);
+  int32_t offset = (int32_t) (limit - target);
+  int32_t count = JERRY_MIN (distance, offset);
+  count >>= shift;
+
+  if (target >= limit || start >= limit || end == 0)
+  {
+    return ecma_copy_value (this_arg);
+  }
+  else
+  {
+    memmove (typedarray_buffer_p + target, typedarray_buffer_p + start, (size_t) (count * element_size));
```

Comment:
If I understand the flow correctly then the following steps are performed:
* the `start/end` values  are multiplied by the `element_size`
* the `end-start` is calculated to get the `count`
* the `count` is divided by the `element_size`
* and lastly: the `count` is multiplied by the `element_size`

Is that a correct observation? If so do we really need to do that many multiplication/division?

Reply:
We need the multiplication with the element_size because we use memmove based on the offset from the start of the buffer, but the other solution is that we use the multiplication directly in the memmove method when we add the offsets like this: typedarray_buffer_p + (target * element_size), i rewrite the code to this, then update the PR, and im looking forward for the upcoming reviews.